mlir/lib/Catalyst/Transforms/DetensorizeFunctionBoundaryPass.cpp

```diff
@@ -0,0 +1,309 @@
+#define DEBUG_TYPE "detensorize-func-boundary"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "Catalyst/IR/CatalystDialect.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst;
+
+namespace {
+bool isScalarTensor(Type type)
+{
+    if (auto rankedType = dyn_cast<RankedTensorType>(type)) {
+        return rankedType.getRank() == 0;
+    }
+    return false;
+}
+
+Type getScalarType(Type type)
+{
+    if (isScalarTensor(type)) {
+        return dyn_cast<RankedTensorType>(type).getElementType();
+    }
+    else {
+        return type;
+    }
+}
+
+bool hasScalarTensorArguments(func::FuncOp funcOp)
+{
+    for (Type type : funcOp.getFunctionType().getInputs()) {
+        if (isScalarTensor(type)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool hasScalarTensorResults(func::FuncOp funcOp)
+{
+    for (Type type : funcOp.getFunctionType().getResults()) {
+        if (isScalarTensor(type)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+struct DetensorizeFuncPattern : public OpRewritePattern<func::FuncOp> {
+    using OpRewritePattern<func::FuncOp>::OpRewritePattern;
+
+    LogicalResult matchAndRewrite(func::FuncOp funcOp, PatternRewriter &rewriter) const override
+    {
+        // Skip for main function
+        if (funcOp->hasAttr("llvm.emit_c_interface")) {
+            return failure();
+        }
+
+        if (!hasScalarTensorArguments(funcOp) && !hasScalarTensorResults(funcOp)) {
+            return failure();
+        }
+
+        // Collect the argument and result types
+        FunctionType funcType = funcOp.getFunctionType();
+        SmallVector<Type> newArgTypes;
+        SmallVector<Type> newResultTypes;
+
+        for (Type type : funcType.getInputs()) {
+            newArgTypes.push_back(getScalarType(type));
```

Comment:
Small suggestion, but something like this could be more readable? 
```suggestion
            newArgTypes.push_back(getScalarOrOriginalType(type));
```

Reply:
updated!